README.md
QtFillingProgressBar
====================

About
-----

The progress bar widget for the Qt toolkit.

Website: https://github.com/AlekseyDurachenko/QtFillingProgressBar

Copyright (c) 2013-2014 Aleksey Durachenko (unless explicitly stated otherwise).

<p align=center>
<img src=doc/examples/statusbar/img/screenshot001.png>
</p>

Usage
-----

```cpp
...
#include <QtFillingProgressBar>
...
CMainWindow::CMainWindow(QWidget *parent) :
    QMainWindow(parent)
{
...
m_progressBar = new QtFillingProgressBar(this);
m_progressBar->setRange(0.0, 1000.0);
m_progressBar->setFormat("Done %p%");
m_progressBar->setPrecision(2);
statusBar()->addWidget(m_progressBar, 1);
...
}
...
```

License
-------

<a href="LICENSE">LGPLv2.1+</a>.


CHANGELOG
v0.1.0
------
first version


src/qtfillingprogressbar.h
// Copyright (C) 2013-2014, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#ifndef QTFILLINGPROGRESSBAR_H
#define QTFILLINGPROGRESSBAR_H

#include <QWidget>

class QtFillingProgressBar : public QWidget
{
    Q_OBJECT
public:
    explicit QtFillingProgressBar(QWidget *parent = 0);
    inline double maximum() const;
    inline double minimum() const;
    inline double value() const;

    inline const QFont &percentFont() const;
    void setPercentFont(const QFont &percentFont);

    inline const QColor &percentColor() const;
    void setPercentColor(const QColor &percentColor);

    inline const QColor &progressColor() const;
    void setProgressColor(const QColor &progressColor);

    inline const QColor &backgroundColor() const;
    void setBackgroundColor(const QColor &backgroundColor);

    inline const QString &format() const;
    void setFormat(const QString &format);

    inline int precision() const;
    void setPrecision(int precision);
protected:
    void paintEvent(QPaintEvent *paintEvent);
    void resizeEvent(QResizeEvent *resizeEvent);
public slots:
    void setRange(double min, double max);
    void setValue(double value);
    void reset();
private:
    double m_min, m_max, m_value;
    QFont m_percentFont;
    QColor m_percentColor;
    QColor m_progressColor;
    QColor m_backgroundColor;
    QString m_format;
    int m_precision;
};

double QtFillingProgressBar::maximum() const
{
    return m_max;
}

double QtFillingProgressBar::minimum() const
{
    return m_min;
}

double QtFillingProgressBar::value() const
{
    return m_value;
}

const QFont &QtFillingProgressBar::percentFont() const
{
    return m_percentFont;
}

const QColor &QtFillingProgressBar::percentColor() const
{
    return m_percentColor;
}

const QColor &QtFillingProgressBar::progressColor() const
{
    return m_progressColor;
}

const QColor &QtFillingProgressBar::backgroundColor() const
{
    return m_backgroundColor;
}

const QString &QtFillingProgressBar::format() const
{
    return m_format;
}

int QtFillingProgressBar::precision() const
{
    return m_precision;
}

#endif // QTFILLINGPROGRESSBAR_H


src/qtfillingprogressbar.cpp
// Copyright (C) 2013-2014, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#include "qtfillingprogressbar.h"
#include <QPaintEvent>
#include <QPainter>

QtFillingProgressBar::QtFillingProgressBar(QWidget *parent) : QWidget(parent)
{
    m_max = 100.0;
    m_min = 0.0;
    m_value = 0.0;
    m_progressColor = Qt::darkBlue;
    m_backgroundColor = Qt::gray;
    m_percentColor = Qt::white;
    m_format = "%p%";
    m_precision = 1;
}

void QtFillingProgressBar::setPercentColor(const QColor &percentColor)
{
    if (m_percentColor != percentColor)
    {
        m_percentColor = percentColor;
        emit update();
    }
}

void QtFillingProgressBar::setPercentFont(const QFont &percentFont)
{
    if (m_percentFont != percentFont)
    {
        m_percentFont = percentFont;
        emit update();
    }
}

void QtFillingProgressBar::setBackgroundColor(const QColor &backgroundColor)
{
    if(m_backgroundColor != backgroundColor)
    {
        m_backgroundColor = backgroundColor;
        emit update();
    }
}

void QtFillingProgressBar::setProgressColor(const QColor &progressColor)
{
    if (m_progressColor != progressColor)
    {
        m_progressColor = progressColor;
        emit update();
    }
}

void QtFillingProgressBar::setFormat(const QString &format)
{
    if (m_format != format)
    {
        m_format = format;
        emit update();
    }
}

void QtFillingProgressBar::setPrecision(int precision)
{
    if (m_precision != precision)
    {
        m_precision = precision;
        emit update();
    }
}

void QtFillingProgressBar::paintEvent(QPaintEvent *)
{
    if (height() > 0 && width() > 0)
    {
        int w = qRound(width()*(m_value-m_min)/(m_max-m_min));
        double percent = 100*(m_value-m_min)/(m_max-m_min);
        QString percentText = m_format;
        percentText.replace("%p", QString::number(percent, 'f', m_precision));

        QPainter painter(this);
        painter.setPen(m_progressColor);
        painter.setBrush(m_progressColor);
        painter.drawRect(0, 0, w, height());
        painter.setPen(m_backgroundColor);
        painter.setBrush(m_backgroundColor);
        painter.drawRect(w, 0, width() - w, height());
        painter.setFont(m_percentFont);
        painter.setPen(m_percentColor);
        painter.drawText(rect(), Qt::AlignCenter, percentText);
    }
}

void QtFillingProgressBar::resizeEvent(QResizeEvent *)
{
    if (m_percentFont.pixelSize() != height())
        m_percentFont.setPixelSize(height());
}

void QtFillingProgressBar::setRange(double min, double max)
{
    m_min = min;
    m_max = max;

    update();
}

void QtFillingProgressBar::setValue(double value)
{
    if (value > m_max)
        m_value = m_max;
    else if (m_value < m_min)
        m_value = m_min;
    else
        m_value = value;

    update();
}

void QtFillingProgressBar::reset()
{
    m_value = m_min;

    update();
}


examples/statusbar/src/main.cpp
// Copyright (C) 2014, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#include "cmainwindow.h"
#include <QApplication>

int main(int argc, char **argv)
{
    QApplication app(argc, argv);

    CMainWindow mainWindow;
    mainWindow.resize(600, 400);
    mainWindow.show();

    return app.exec();
}


examples/statusbar/src/cmainwindow.h
// Copyright (C) 2014, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#ifndef CMAINWINDOW_H
#define CMAINWINDOW_H

#include <QtFillingProgressBar>
#include <QMainWindow>
#include <QTimer>


class CMainWindow : public QMainWindow
{
    Q_OBJECT
public:
    explicit CMainWindow(QWidget *parent = 0);
signals:
private slots:
    void onStartProgressBar();
    void onStopProgressBar();
    void onTimerTimeOut();
private:
    QTimer *m_timer;
    QtFillingProgressBar *m_progressBar;
};


#endif // CMAINWINDOW_H


examples/statusbar/src/cmainwindow.cpp
// Copyright (C) 2014, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#include "cmainwindow.h"
#include <QStatusBar>
#include <QMenuBar>
#include <QLayout>
#include <QPushButton>

CMainWindow::CMainWindow(QWidget *parent) :
    QMainWindow(parent)
{
    setWindowTitle("The progress bar example");
    setStatusBar(new QStatusBar(this));
    setMenuBar(new QMenuBar(this));
    QMenu *fileMenu = menuBar()->addMenu("File");
    QAction *startAction = fileMenu->addAction("&Start progress bar");
    QAction *stopAction = fileMenu->addAction("&Stop progress bar");
    fileMenu->addSeparator();
    QAction *quitAction = fileMenu->addAction("&Quit");
    quitAction->setStatusTip("Quit");
    connect(startAction, SIGNAL(triggered()), this, SLOT(onStartProgressBar()));
    connect(stopAction, SIGNAL(triggered()), this, SLOT(onStopProgressBar()));
    connect(quitAction, SIGNAL(triggered()), this, SLOT(close()));

    QWidget *centralWidget = new QWidget(this);
    centralWidget->setLayout(new QHBoxLayout(centralWidget));
    QPushButton *startButton = new QPushButton("Start progress bar", this);
    QPushButton *stopButton = new QPushButton("Stop progress bar", this);
    centralWidget->layout()->addWidget(startButton);
    centralWidget->layout()->addWidget(stopButton);
    setCentralWidget(centralWidget);
    connect(startButton, SIGNAL(clicked()), this, SLOT(onStartProgressBar()));
    connect(stopButton, SIGNAL(clicked()), this, SLOT(onStopProgressBar()));

    m_timer = new QTimer(this);
    m_timer->setInterval(10);
    connect(m_timer, SIGNAL(timeout()), this, SLOT(onTimerTimeOut()));

    m_progressBar = new QtFillingProgressBar(this);
    m_progressBar->setRange(0.0, 1000.0);
    m_progressBar->setFormat("Done %p%");
    m_progressBar->setPrecision(2);
    statusBar()->addWidget(m_progressBar, 1);
    m_progressBar->hide();
}

void CMainWindow::onStartProgressBar()
{
    m_progressBar->show();
    m_progressBar->reset();
    m_timer->start();
}

void CMainWindow::onStopProgressBar()
{
    m_progressBar->hide();
    m_timer->stop();
}

void CMainWindow::onTimerTimeOut()
{
    m_progressBar->setValue(m_progressBar->value() + 1.0);
    if (m_progressBar->value() >= 1000.0)
        onStopProgressBar();
}
